### PR TITLE
fix: add status code if the body contain a code

### PR DESCRIPTION
**Pull request by ricantar**

### CODE REVIEW

File: provider/provider.go [hunk 1]
```diff
@@ -785,42 +786,64 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
+		return nil, &RelayOutputErr{Error: parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey), StatusCode: rawOutput.Status}
 	}
 
 	return parseRelaySuccesfulOutput(bodyBytes)
 }
 
-func parseRelaySuccesfulOutput(bodyBytes []byte) (*RelayOutput, error) {
+// TODO: REMOVE THIS HACKY FUNCTION
+func extractStatusFromResponse(response string) string {
+	patterns := []string{
```
Comment (Olshansk) [inline provider/provider.go:817]: I'd precompile these.

```go
// Pre-compile the regular expressions for efficiency
var regexPatterns = []*regexp.Regexp{
	regexp.MustCompile(`"code"\s*:\s*(\d+)`),       // Matches `"code": 4XX` or similar if came from the node
	regexp.MustCompile(`(\d+)\s+Not Found`),        // Matches `404 Not Found` or similar
	regexp.MustCompile(`(\d+)\s+page not found`),   // Additional pattern for matching `404 page not found`
	regexp.MustCompile(`HTTP\/\d\.\d\s+(\d+)`),     // Matches `HTTP/1.1 4XX` or similar, for HTTP responses
	regexp.MustCompile(`"statusCode"\s*:\s*(\d+)`), // Matches `"statusCode": 4XX` or similar
}
```

Comment (Olshansk) [reply provider/provider.go:817]: You can move it to an `init()` function in this file.

File: provider/provider.go [hunk 2]
```diff
@@ -785,42 +786,64 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
 	if errors.Is(reqErr, errOnRelayRequest) {
-		return nil, parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey)
+		return nil, &RelayOutputErr{Error: parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey), StatusCode: rawOutput.Status}
 	}
 
 	return parseRelaySuccesfulOutput(bodyBytes)
 }
 
-func parseRelaySuccesfulOutput(bodyBytes []byte) (*RelayOutput, error) {
+// TODO: REMOVE THIS HACKY FUNCTION
```
Comment (Olshansk) [inline provider/provider.go:815]: Is this an actual TODO? 

1. If so, can you leave a comment on how you're thinking about removing it and why?
2. If not, add a godoc explaining what this function is.

I'm personally unsure (off the top off my head) what the better solution is.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Couple small nits.

File: provider/provider.go (hunk 1 above)
Comment (ricantar) [reply provider/provider.go:817]: I don't think we want to do that for now. This is a temporary function that will only be called when used the relay() function, and we have more than one function and many other cases were I don't see why we should load the regex

File: provider/provider.go (hunk 2 above)
Comment (ricantar) [reply provider/provider.go:815]: I can update the comment. The idea is that the node respond back with the status for REST chains, so the node must get back to us with response, signature and statusCode

File: provider/provider.go (hunk 1 above)
Comment (ricantar) [reply provider/provider.go:817]: But Relay is our must used function here, I think it could be worth

File: provider/provider.go (hunk 2 above)
Comment (Olshansk) [reply provider/provider.go:815]: Providing this level of context in the comment is perfect!

File: provider/provider.go (hunk 1 above)
Comment (Olshansk) [reply provider/provider.go:817]: There's no downside as far I can see.

Either you (regardless of whether its used in other place):

1. Recompile regexes on every call
2. Do it once

File: provider/provider.go [hunk 3]
```diff
@@ -785,42 +786,63 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
 	if errors.Is(reqErr, errOnRelayRequest) {
-		return nil, parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey)
+		return nil, &RelayOutputErr{Error: parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey), StatusCode: rawOutput.Status}
 	}
 
 	return parseRelaySuccesfulOutput(bodyBytes)
 }
 
-func parseRelaySuccesfulOutput(bodyBytes []byte) (*RelayOutput, error) {
+// TODO: Remove this function after the node respond back to us with a statusCode along side with the response and the signature
```
Comment (Olshansk) [inline provider/provider.go:815]:
```suggestion
// TODO: Remove this function after the node responds back to us with a statusCode alongside with the response and the signature.
// Returns "200" if none of the pre-defined internal regexes matches any return values.
```

Comment (Olshansk) [inline provider/provider.go:815]: Questions: Will this require `pocket-core` changes?

Not a loaded question. Just want to know what you have in mind so we can plan / design properly for Shannon.

File: provider/provider.go [hunk 4]
```diff
@@ -785,42 +786,63 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
+func extractStatusFromResponse(response string) string {
+	patterns := []*regexp.Regexp{
```
Comment (Olshansk) [inline provider/provider.go:817]: NIT: I still think this should be global and initialized once in an `init` but not going to block.

CC @adshmh for another POV

File: provider/provider.go [hunk 5]
```diff
@@ -785,42 +786,63 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
+		regexp.MustCompile(`"statusCode"\s*:\s*(\d+)`), // Matches `"statusCode": 4XX` or similar
+	}
+
+	for _, pattern := range patterns {
+		matches := pattern.FindStringSubmatch(response)
+		if len(matches) > 1 {
+			return matches[1] // Return the first captured group, which should be the status code.
+		}
+	}
+	return "200"
```
Comment (Olshansk) [inline provider/provider.go:831]: Are you sure you want to default to a success if none of the error cases we account for are matched?

I'm thinking one of two non mutually-exclusive ideas:

1. Add a regex check for `200` success
2. Update the method header to `func extractStatusFromResponse(response string) (statusCode string, known bool)`

And default to `"unknown", false` as a default case signifying "def not an error" but also "relaying logic elsewhere on success`

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: provider/provider.go (hunk 3 above)
Comment (ricantar) [reply provider/provider.go]: I think we already discuss this in a previous call (if not, here's the TL;DR: The node respond must come with signature, response and statusCode), so yes, this should be implemented in Shannon

File: provider/provider.go (hunk 5 above)
Comment (ricantar) [reply provider/provider.go:832]: Yes, we want the default to be 200 because of normal chains (which should always get back to us with 200). Even tho it's really valuable to get those 200 cases, for now everything is "unknown" because the node are getting back to us with 200 no matter what

Comment (Olshansk) [reply provider/provider.go:832]: Not going to block this because you're the primary user of the library so ultimately it's up to you, but I do still thinking that defaulting to success it not a good design decision. Defaulting to unknown (can be a constant in the SDK) and handling that appropriately prevents requiring "implicit knowledge" that "200" may not necessarily a "200".

[1] https://chat.openai.com/share/111342f3-0045-4e71-b24b-4e2b8822dd25

Side note: ChatGPT also suggested extracting the regex's outside of the function.

Comment (Olshansk) [reply provider/provider.go:832]: @adshmh Bump on this one too. Need a 3rd data point.

Comment (ricantar) [reply provider/provider.go:832]: The regex it's already in a separate init, will be deleted later but for now can be there. But the default to unknown it's exactly what we have now and that's what we want to avoid. However since this is a temporary patch it's not really our intended design

Comment (Olshansk) [reply provider/provider.go:832]: Wdyt: #94

Comment (ricantar) [reply provider/provider.go:832]: Looks good. But I made this change already, it's not pushed bc I'm testing the integration with the gateway 
image
image


I'll push this version on monday

Comment (ricantar) [reply provider/provider.go:832]: But I liked that last comment, I'm copying it 😛, it's way better to have it clear in the code than depending on my memory/knowledge: 

// NOTE_HACK: Please note that if no error is found (per the list of regexes)
	// above, we default to a successful status code of "200" instead of an "unknown".
	// This is done intentionally to achieve parity with the prior design until
	// a correct long-term solution is implemented.

Comment (Olshansk) [reply provider/provider.go:832]: >  it's way better to have it clear in the code than depending on my memory/knowledge:

💯 

It's how we can scale development and engineering teams:
- Docs are nice, but how often do engineers really go read them?
- Memory / institutional knowledge is important, but what happens when the primary POC  is OOO?
- We have conversations/discussions in meetings/code review comments, but it's the same amount of effort to document it in the code, and that's usually the first place we go look to.

---

tl;dr Copy as much as you wish 😉

File: provider/provider.go (hunk 4 above)
Comment (adshmh) [reply provider/provider.go:818]: Yes, I agree. These should only be constructed once: all of these regexps are knowin in advance, and  compiling them as part of processing every request is unnecessary (no strong opinions here on where the one-time initialization should take place)

File: provider/provider.go (hunk 5 above)
Comment (adshmh) [reply provider/provider.go:832]: Regarding `unknown` vs. 200 , I think it is best to have "unknown" as the default return value, and return 200 only if it is explicitly identified in the core's response. Let the user of the library decide on what `unknown` means.

This makes the most sense to me because it makes sure the library does not mask any key detail: actual 200 vs. unknown is an important detail.
Although it is correctly pointed out that  ```for now everything is "unknown" because the node are getting back to us with 200 no matter what```, this library should not build on this missing/incorrect behavior and further complicate matters downstream. It should instead defer this decision to the user by returning all the available data pieces, which also removes the need for knowledge on the history of the matter when reading the library's code.

Also, "temporary" solutions may stay in-place for much longer than expected for a variety of reasons, and therefore still deserve the general application of best practices IMO.

Comment (ricantar) [reply provider/provider.go:832]: I have to disagree, the default status code for an HTTP call is 200 (which means OK I got your call), unknown is not a valid status code and should not be used. Here the definition of the RC 9110. 
https://httpwg.org/specs/rfc9112.html#status.line 
https://httpwg.org/specs/rfc9110.html#overview.of.status.codes

A status code must have 3 digit and we have to define a default ourself in the library (i.e: GO default is 200, if you forget to set one it will be set to 200 automatically)

Comment (Olshansk) [reply provider/provider.go:832]: _tl;dr ASSUMING we are building a long-term solution in this PR, I suggest defeaulting to `202` and having explicit checks for `200`._

---

I did not have strong opinions about this under the assumption that it was a hack / temp / workaround in `Morse`, but given @ricantar messages that the same default will apply to `Shannon` (see the image below), will provide another argument.

Screenshot 2024-05-06 at 10 36 57 AM

- Regarding having a `3 digist status code` instead of `unknown` -> Agreed.
- Regarding using `200` -> Disagreed.

After asking perplexity (see search [here](https://www.perplexity.ai/search/What-is-the-S3wPFB55SGGnnn6oYZn.sQ)), I looked into `202` and `204` and believe `202` may be the most appropriate, and least misleading, from the POV of an SDK..

---

[What is the difference between a 200 status code and a 202 status code?](https://www.webfx.com/web-development/glossary/http-status-codes/what-is-a-200-status-code/):

> When a server returns a 200 OK status code, it means that the server has successfully processed the client’s request and is returning the requested content in the response body.

- 202: The request went through but we don't know what happened
- 200: Explicit SUCCESS / OK

Comment (ricantar) [reply provider/provider.go:832]: I can agree with that, 202 it's better for a default

File: provider/provider.go (hunk 4 above)
Comment (ricantar) [reply provider/provider.go]: Done

File: provider/provider.go (hunk 5 above)
Comment (ricantar) [reply provider/provider.go]: Comments addressed

File: provider/provider.go [hunk 6]
```diff
@@ -30,6 +33,27 @@ var (
...
+		"tls: handshake failure":              525, // SSL Handshake Failed (Cloudflare specific, non-standard)
+		"i/o timeout":                         504, // Gateway Timeout
+		"bad gateway":                         502, // Bad Gateway
+		"service unavailable":                 503, // Service Unavailable
+		"gateway timeout":                     504, // Gateway Timeout
+	}
+)
+
+const (
+	defaultCode = 202
```
Comment (Olshansk) [inline provider/provider.go:55]: 1. NIT: Can you please add an explanation above `202` as to why this is the default status code? If feeling to lazy, just copy-paste a link to https://github.com/pokt-foundation/pocket-go/pull/93#discussion_r1586616825.
2. `s/defaultCode/defaultStatusCode/g` to match the name of `errorStatusMap` above

File: provider/provider.go [hunk 7]
```diff
@@ -30,6 +33,27 @@ var (
...
+	resultText  = "result"
```
Comment (Olshansk) [inline provider/provider.go:56]: Can you add a comment explaining the goal / point / usage of this constant?

File: provider/provider.go [hunk 8]
```diff
@@ -785,42 +821,92 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
-		return nil, err
+		return nil, &RelayOutputErr{Error: err, StatusCode: statusCode}
 	}
 
 	if errors.Is(reqErr, errOnRelayRequest) {
-		return nil, parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey)
+		return nil, &RelayOutputErr{Error: parseRelayErrorOutput(bodyBytes, input.Proof.ServicerPubKey), StatusCode: statusCode}
 	}
 
+	// The statusCode will be overwrrite based on the response
```
Comment (Olshansk) [inline provider/provider.go:849]:
```suggestion
	// The statusCode will be overwritten based on the response
```

File: provider/provider.go [hunk 9]
```diff
@@ -785,42 +821,92 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
+	statusCode := defaultCode
+
+	if reqErr != nil {
+		for key, status := range errorStatusMap {
+			if strings.Contains(reqErr.Error(), key) { // This checks if the actual error contains the key string
+				return status
+			}
+		}
+
+		// If we got an error and we can't identify it as a known error, will be mark as if the server failed
```
Comment (Olshansk) [inline provider/provider.go:863]: Great comments in the implementation

File: provider/provider.go [hunk 10]
```diff
@@ -785,42 +821,92 @@ func (p *Provider) DispatchWithCtx(ctx context.Context, appPublicKey, chain stri
...
+		return http.StatusInternalServerError
+	}
+
+	if rawOutput.StatusCode != http.StatusOK {
```
Comment (Olshansk) [inline provider/provider.go:867]: Love the usage of whitespace and how easy this is to read!

**Review by Olshansk — DISMISSED** [review]
Great work @ricantar and I love the shape this has taken.

I left a few NITs ONLY related to comments.

PTAL and update before merging, but otherwise :shipit:

File: provider/provider.go (hunk 8 above)
Comment (ricantar) [reply provider/provider.go]: Done

File: provider/provider.go (hunk 7 above)
Comment (ricantar) [reply provider/provider.go]: Did it

File: provider/provider.go (hunk 6 above)
Comment (ricantar) [reply provider/provider.go]: Did it too

**Review by Olshansk — APPROVED** [review]
Please make sure to update the git commit message when you squash & mege.
